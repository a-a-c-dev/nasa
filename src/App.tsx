import {Suspense,lazy } from 'react';
import {Spinner} from './components/index';
import {BrowserRouter, Routes, Route} from 'react-router-dom';
const MarsRover = lazy (()=>import('./pages/MarsRover')) ;
const PictureOfTheDay = lazy (()=>import('./pages/PictureOfTheDay')) ;
const ApolloLaunch = lazy (()=>import('./pages/ApolloLaunch')) ;
const NearEarthObjects = lazy (()=>import('./pages/NearEarthObjects')) ;
const NasaProjects = lazy (()=>import('./pages/NasaProjects')) ;
import ErrorBoundary from './components/ErrorBoundary';

function App() {
  return (
    <>
      <ErrorBoundary>
        <BrowserRouter>
          <Routes>
            <Route index element={     
              <Suspense  fallback={<Spinner/>}>
                <PictureOfTheDay />      
              </Suspense>}/>
            <Route path='MarsRover' element={
              <Suspense  fallback={<Spinner/>}>
                <MarsRover />      
              </Suspense>} />
            <Route path='ApolloLaunch' element={
              <Suspense  fallback={<Spinner/>}>
                <ApolloLaunch />      
              </Suspense>} />
            <Route path='NearEarthObjects' element={
              <Suspense fallback={<Spinner/>}>
                <NearEarthObjects/>
              </Suspense>}/>
            <Route path='projects' element={
              <Suspense fallback={<Spinner/>}>
                <NasaProjects/>
              </Suspense>}/>
          </Routes>
        </BrowserRouter>
      </ErrorBoundary>
    </>
  )
}

export default App