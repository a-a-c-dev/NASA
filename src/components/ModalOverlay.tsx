
import React,{useState, ReactNode} from 'react'
import  ReactDOM from 'react-dom';

interface ModalOverlayProps{
  toggle:(show: ()=> void) => ReactNode,
  content:(hide: ()=> void) => ReactNode
}

const ModalOverlay = React.memo(({ toggle, content }:ModalOverlayProps) => {
    const [isShown, setIsShown] = useState(false);
    const hide = () => setIsShown(false);
    const show = () => setIsShown(true);
    return (
      <>
        {toggle(show)}
        {isShown && content(hide)}
      </>
    );
  })
  
  export const Modal = ({ children}:{children: ReactNode}) => {
    
   const targetElement = document.getElementById('modal-root');
   if(!targetElement)return null

   return ReactDOM.createPortal(
      <div className="modal-overlay">
        <div className="modal">
          {children}
        </div>
      </div>
      ,
    targetElement
    
  );

}


export default ModalOverlay;