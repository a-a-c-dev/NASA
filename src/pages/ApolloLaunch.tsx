import useFetch from '../hooks/useFetch';
import {Header,Navigation,ImageContainer,Error,Spinner, Modal, ModalOverlay} from '../components/index';


let url= "https://images-api.nasa.gov/search?q=apollo+11...&page=1"

interface Fetch{
  data: Images[],
  loading: boolean| null,
  error : string | null
}

interface Images{
  links:{
    0:{
      href:string
    }
  }
  data:{
    0:{
      date_created:string,
      title:string,
      description:string
    }
    
  }
  
} 

function ApolloLaunch() {
  const {data,error, loading} = useFetch<Fetch>(url, 'Apollo-Launch')
  const images:Images[] = data ? Object.values(data)[0].items.slice(0,20) : []
  
  if(loading){
    return(
      <>
        <Navigation/>
        <Header text='Apollo Launch'/>
          <p>please wait while the data is loading</p>
        <Spinner/>
      </>
    )
  }
  
  if(error)return <Error error={error}/>

  return (   
    <>
    <Navigation/>
    <Header text='Apollo Launch'/>
    <div className='main-container apollo'>  
            {images && images?.map((value:Images) => (
                value?.links?(
                         <ModalOverlay
                          key={value.data[0].title}
                          toggle={show =>         
                          <div className='card' onDoubleClick={()=> show()} >
                              <ImageContainer imageSrc={value?.links[0]?.href} imagedesc={value?.data[0]?.description}/>
                              <div className='info-container'>
                                <h3>date: {value?.data[0]?.date_created.slice(0,10)}</h3>
                                <h2>title: {value?.data[0]?.title.substr(0,50)}</h2>
                                <p>description: {value?.data[0]?.description.substr(0,100)}</p>
                              </div>
                              <span>double click to see more</span>

                          </div>
                          }
                          content={hide => (
                                <Modal>
                                  <div className='card'> 
                                      <button className='close-btn' onClick={hide}>X</button>
                                      <ImageContainer imageSrc={value?.links[0]?.href} imagedesc={value?.data[0]?.description}/>
                                      <div className='info-container'>
                                        <h3>date: {value?.data[0]?.date_created.slice(0,10)}</h3>
                                        <h2>title: {value.data[0]?.title}</h2>
                                        <p>description: {value?.data[0]?.description}</p>
                                      </div>                    
                                  </div>
                                    
                               </Modal>
                          )}
                    />
                                
                ): null
                
            ))}
          
    </div>
    </>
     
  )
}

export default ApolloLaunch


