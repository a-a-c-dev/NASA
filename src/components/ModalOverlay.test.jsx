import { describe, expect, test } from 'vitest';
import { render, screen,fireEvent, waitFor ,act} from '@testing-library/react';
import ModalOverlay, { Modal } from './ModalOverlay';
import ApolloLaunch from '../pages/ApolloLaunch';
import  ReactDOM from 'react-dom';
import App from '../App'
import { BrowserRouter,MemoryRouter  } from 'react-router-dom';

describe('Modal component', () => {
    const hide = vi.fn()
    const show = vi.fn()
    test('The component should render in body when open', () => {
        const {baseElement} = render(
            <ModalOverlay toggle={show}>
                <Modal hide={hide}/>
            </ModalOverlay>
        );
        expect(baseElement).toMatchSnapshot();
    });
  
});
