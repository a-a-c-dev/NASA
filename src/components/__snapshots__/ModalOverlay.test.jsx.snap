// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal component > The component should render in body when open 1`] = `
<body>
  <div />
</body>
`;
